src/ray/object_manager/pull_manager.cc

```diff
@@ -96,30 +98,37 @@ void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
     return;
   }
 
+  // New object locations were found, so begin trying to pull from a
+  // client. This will be called every time a new client location
+  // appears.
+  bool did_pull = PullFromRandomLocation(object_id);
+  if (did_pull) {
+    UpdateRetryTimer(request);
+    return;
+  }
+
+  // If we cannot pull, try restoring objects from the external storage.
   if (!request.spilled_url.empty()) {
-    // Try to restore the spilled object.
+    const auto spilled_node_id = request.spilled_node_id;
+    RAY_CHECK(!spilled_node_id.IsNil());
+    // Try to restore the spilled object from a local or remote node.
     restore_spilled_object_(
-        object_id, request.spilled_url, [this, object_id](const ray::Status &status) {
-          bool did_pull = true;
-          // Fall back to fetching from another object manager.
+        object_id, spilled_node_id,
+        [object_id, spilled_node_id](const ray::Status &status) {
           if (!status.ok()) {
-            did_pull = PullFromRandomLocation(object_id);
-          }
-          if (!did_pull) {
-            RAY_LOG(WARNING) << "Object restoration failed and the object could not be "
-                                "found on any other nodes. Object id: "
-                             << object_id;
+            RAY_LOG(WARNING)
+                << "Object restoration request to a remote failed and the object could "
+                   "not be "
+                   "found on any other nodes. It could be because the remote node that "
+                   "spilled the object has crashed or the network request has failed. "
+                   "Please check the log of node of id: "
+                << spilled_node_id << " Object id: " << object_id;
           }
         });
+    // Q: Maybe we shouldn't update the timer and rate limit the restore requests within a
```

Comment:
cc @stephanie-wang @ericl This can be a potential place for performance degradation. would love to hear suggestions?

Reply:
I think we should probably just never update the timer and make the RPC to be sent every 1 second or something at max.

Reply:
I think it's premature to try to optimize this since we haven't seen this situation yet. It probably makes sense to reset the timer if the restore failed, though (#13514 adds a similar reset when a needed object is evicted). Then, it should be safe to try again since we shouldn't have any other pull/restores in flight.

Reply:
I think this can cause some noticeable delay in some scenarios though. For example, here's a typical object restoration path

```
   node A                         node B
1. restore request  =>
2.                                      restore_object from external storage
3.                                      update the location to GCS
4. message published
5. Send a pull request => 
```
and if this path takes a long time (which can happen in some scenarios since we have two slow paths, restore_object, and GCS location update), the delay can be pretty significant since we are retrying with exponential backoff. For example, if this delays more than 700ms, the delay could be as long as a half-second (10+20+40+80+160+320+640) whereas if we handle this properly, it'd take only 10ms. 

But I agree this type of scenario is not seen in real life yet. What about this? I will try benchmarking the multi-node spilling performance (with cluster utils) against the multi-node file spilling V1 PR (which won't have this issue), and if the performance is not that different, I will just comment about it. 

Also, I will follow your suggestion about resetting in upon failures.

Reply:
I agree with a fast retry here (10ms?). The only reason we had the exponential backoff previously was to avoid duplicate pull requests. It seems like this isn't an issue since we de-dup restores and restore requests all go to the same node right?

Reply:
Yeah. This can still kind of spamming the node though. There will be 100 * (number of nodes requesting restoration) / S requests to the node that spilled the object if we don't have some logic to suppress requests.

Btw, I just verified without removing this, the streaming shuffle (4 nodes) doesn't seem to work.